### PR TITLE
fix(qwen3): chunk echo prefill to bound prompt-logprob activation scratch

### PR DESCRIPTION
**Pull request by wjinxu**
## Summary
  
Echo with `logprobs` returns a logprob for every prompt token, which the old path produced by materializing all-position logits (`vocab × prompt_len`) in a single forward. Echo was exempt from prefill chunking, so a long echo prompt's activation scratch was never budgeted — a well-formed request (e.g. ~12k-token prompt) deterministically OOMed *after* admission instead of being served or rejected up front.

This chunks echo prefill like any other prompt: each chunk computes all-position logits only for its own slice (`vocab × chunk`) and the executor stitches the per-chunk prompt logprobs back together, so the scratch is bounded by the prefill budget regardless of prompt length.

## Changes

- **scheduler** (`take_prefill_chunks`): drop the echo exemption; echo chunks under the shared prefill budget like everything else.
- **executor**: build a per-chunk *partial* of prompt logprobs (only the positions this chunk forwarded are `Some`) and fold it into a per-request accumulator, surfacing the full vector only on the final chunk. The accumulator is cleared on `drop_request` so a mid-prefill disconnect can't leak.
- **plan**: route `echo + logprobs` to a dedicated Prefill step, since the unified forward can't produce all-position logits (would silently drop the echoed prompt logprobs). Also gate the all-position-logits scratch on `echo + logprobs`, so echo *without* logprobs (which only echoes prompt ids back) no longer pays for that scratch.

## Correctness

The seam between chunks is the delicate part: prompt token `k`'s logprob comes from the distribution at position `k-1`, so a boundary token's logprob is produced by the *previous* chunk's last position. Each prompt index is produced by exactly one chunk, so folding partials never drops or double-writes a slot.

## Tests

- Unit: chunk-fold seam reconstruction, echo-chunks-under-budget scheduling, and the plan gating (`echo+logprobs` → dedicated prefill; all-position-logits fire only for `echo+logprobs`).
- GPU self-consistency gate (`tests/echo_chunked_prefill.rs`): chunked echo prompt logprobs must match the single-pass result within bf16 tolerance (no HF golden for within-prompt distributions, so this asserts a hardware-independent invariant instead). Skips cleanly without the model.

## Tradeoff

While an `echo + logprobs` request is mid-prefill it forces dedicated Prefill steps, so active decodes wait those ticks (no interleaving). `echo + logprobs` is rare and this trades a small latency hit for correctness; the previous unified path silently dropped the prompt logprobs.

### CODE REVIEW

File: openinfer-qwen3-4b/src/scheduler/plan.rs [hunk 1]
```diff
@@ -33,11 +33,20 @@ pub(super) enum ExecutionArtifacts {
...
     pending: Vec<PendingRequest>,
 ) -> Option<ExecutionPlan> {
-    if !pending.is_empty() && have_active {
+    // echo+logprobs needs a dedicated Prefill: the unified forward can't produce
+    // all-position logits. Active decodes wait those ticks; it's rare.
+    let needs_dedicated_prefill = batch_needs_prompt_logprobs(&pending);
+    if !pending.is_empty() && have_active && !needs_dedicated_prefill {
         Some(ExecutionPlan::Unified { pending })
     } else if !pending.is_empty() {
         Some(ExecutionPlan::Prefill { pending })
```
[the remark this answers is not in the document]
Comment (wjinxu) [reply openinfer-qwen3-4b/src/scheduler/plan.rs:52]: fixed

**Comment by xiaguan** [comment]
@codex review

**Comment by xiaguan** [comment]
Thanks for digging into this — the chunk-fold seam logic and the failure-target fix are both solid work.

Before we land it I want to make sure we're fixing the reported bug. The #358 repro is a plain `/v1/completions` request — and on that path `echo` is hardcoded `false` (`bridge.rs:240`) and `prompt_logprobs` is rejected outright (`sampling_guard.rs:188`), so the echo + all-position-logits path this PR chunks is never reached by an API request. The ~22 KB/token I measured is the normal last-token-logits prefill, not the `vocab × prompt_len` spike.

Also — and this one is on me, the issue was stale — per-request prefill chunking already landed in #363, so the "prefill scratch is never budgeted" half of #358 is done. What's actually left is admission accounting for single-request activation headroom, which this PR doesn't touch.

Could you confirm whether your repro used `echo`? If it was a plain prompt, this PR doesn't address #358 and we should unlink it. The echo hardening itself is real, but it's currently unreachable from the API (echo off + prompt_logprobs rejected), so I'd rather not carry the extra per-request accumulator + dedicated-prefill routing until echo is actually wired.

I'll tighten #358 to spell out the repro params and the remaining admission gap so the next person isn't sent down the same path.

**Comment by wjinxu** [comment]
Thanks for the careful read — you're right, and I appreciate you catching this before it landed.

  Confirming the repro: it was a plain /v1/completions request, no echo. So this PR is targeting a path that isn't reachable from the API today (echo hardcoded false + prompt_logprobs rejected at the guard), and what I measured was the normal last-token-logits prefill, not the vocab × prompt_len spike I assumed. Combined with #363 already covering per-request prefill chunking, this PR doesn't move #358 forward.

  Closing this out and unlinking from #358. The chunk-fold seam logic and the partial-accumulator design should still be useful whenever echo gets wired through the bridge — happy to revisit then. I'll wait for your updated #358 to see where the real admission-accounting gap lives before picking up anything new here.
